src/builder/commands/pip.rs

```diff
@@ -326,11 +341,18 @@ fn version_req(hash: &mut Digest, fname: &Path) -> Result<(), VersionError> {
                 if chunk.len() == 0 || chunk.starts_with("#") {
                     continue;
                 }
+                if let Some(req) = parse_req_filename(chunk) {
+                    included_files.push(req);
+                }
                 // Should we also ignore the order?
                 hash.item(chunk);
             }
             Ok(())
-    }).map_err(err)
+    }).map_err(err));
+    for req in &included_files {
```

Comment:
Sure, this has nothing to do with the recursion, but with the error reporting.

I.e. you need to turn all errors into VersionError instead of io::Error.

So you move `map_err(err)` after `and_then` to before `and_then`. Then you need to wrap `try!(line)` into the `try!(line.map_err(err))`. Then you can `try!(version_req(...))` without the issues because error type is unified everywhere.